indexer/cmd/indexer.go

```diff
@@ -0,0 +1,319 @@
+				if err != nil {
+					bi.logger.Error("failed to fetch blocks", "start", blockNum, "end", endBlockNum, "error", err)
+					break
+				}
+
+				for _, block := range blocks {
+					bi.backwardBlockChan <- block
+					bi.lastBackwardIndexedBlock.Set(block.Number())
+					if block.Number().Cmp(zeroBigNum) == 0 {
+						bi.logger.Info("done fetching backward blocks...")
+						return
+					}
+				}
+				blockNum.Sub(endBlockNum, big.NewInt(1))
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) processForwardBlocks(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case block := <-bi.forwardBlockChan:
+			if err := bi.indexBlock(ctx, block); err != nil {
+				bi.logger.Error("failed to index block", "error", err)
+			}
+			if err := bi.indexTransactions(ctx, block); err != nil {
+				bi.logger.Error("failed to index transactions", "error", err)
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) processBackwardBlocks(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case block := <-bi.backwardBlockChan:
+			if err := bi.indexBlock(ctx, block); err != nil {
+				bi.logger.Error("failed to index block", "error", err)
+			}
+			if err := bi.indexTransactions(ctx, block); err != nil {
+				bi.logger.Error("failed to index transactions", "error", err)
+			}
+			if block.Number().Cmp(big.NewInt(0)) == 0 {
+				bi.logger.Info("done processing backward blocks...")
+				return
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) indexBlock(ctx context.Context, block *types.Block) error {
+	timestamp := time.UnixMilli(int64(block.Time())).UTC().Format("2006-01-02T15:04:05.000Z")
+	indexBlock := &store.IndexBlock{
+		Number:       block.NumberU64(),
+		Hash:         block.Hash().Hex(),
+		ParentHash:   block.ParentHash().Hex(),
+		Root:         block.Root().Hex(),
+		Nonce:        block.Nonce(),
+		Timestamp:    timestamp,
+		Transactions: len(block.Transactions()),
+		BaseFee:      block.BaseFee().Uint64(),
+		GasLimit:     block.GasLimit(),
+		GasUsed:      block.GasUsed(),
+		Difficulty:   block.Difficulty().Uint64(),
+		ExtraData:    hex.EncodeToString(block.Extra()),
+	}
+
+	return bi.storage.IndexBlock(ctx, indexBlock)
+}
+
+func (bi *BlockchainIndexer) indexTransactions(ctx context.Context, block *types.Block) error {
+	var transactions []*store.IndexTransaction
+	var txHashes []string
+
+	for _, tx := range block.Transactions() {
+		from, err := types.Sender(types.NewCancunSigner(tx.ChainId()), tx)
+		if err != nil {
+			return fmt.Errorf("failed to derive sender: %w", err)
+		}
+
+		v, r, s := tx.RawSignatureValues()
+		timestamp := tx.Time().UTC().Format("2006-01-02T15:04:05.000Z")
+		transaction := &store.IndexTransaction{
+			Hash:        tx.Hash().Hex(),
+			From:        from.Hex(),
+			Gas:         tx.Gas(),
+			Nonce:       tx.Nonce(),
+			BlockHash:   block.Hash().Hex(),
+			BlockNumber: block.NumberU64(),
+			ChainId:     tx.ChainId().String(),
+			V:           v.String(),
```

Comment:
Are there more meaningful names for the `V`, `R`, `S` fields?

Reply:
yes they are part of signatures.